docs/sql/udf.md

```diff
@@ -11,106 +11,266 @@ The following example shows a user-defined function:
 
 :::
 
-The SQL statement `CREATE FUNCTION` can be used to declare new
-functions whose implementation is provided separately in Rust.
+:::danger
+
+* Feldera's incremental query engine assumes that all computations are deterministic. Using
+  a non-deterministic UDF is likely to result in incorrect outputs. The SQL compiler cannot verify
+  that Rust functions meet this requirement. Therefore, it is the responsibility of the UDF
+  developer to ensure that their UDFs are deterministic, i.e., the function's return value
+  must depend solely on its arguments and not on clocks, random number generators, external
+  data sources, or other sources of non-determinism.
+
+  We are considering adding support for non-deterministic UDFs in the future. If your use case
+  requires a non-deterministic UDF, please share your feedback by commenting on the associated
+  [GitHub issue](https://github.com/feldera/feldera/issues/2619)
+
+* UDFs are compiled into native binary code and executed directly within the address space of
+  the pipeline. Therefore, only trusted code should be included in UDFs.  UDFs should not contain
+  panics or invoke undefined behaviors. These can lead to crashes, memory corruption, or
+  unpredictable behavior within the pipeline.  UDFs should handle all runtime errors gracefully,
+  by returning a `Result::Err()` ([see below](#return-types)). This ensures that errors are
+  propagated in a controlled manner and can be handled by the calling code.
+:::
 
-The implementation of these functions must be provided to the Rust
-compiler in a separate file.  Moreover, when the SQL program contains
-user-defined functions, the compiler will generate an additional
-output file named `protos.rs` containing the signature for all the
-Rust user-defined functions.
+Feldera supports UDFs implemented in Rust. To define a UDF, users declare the function
+in SQL without providing a body. The corresponding Rust implementation is supplied in a separate
+Rust file, which contains the logic for all UDFs declared within the SQL program. A complete
+set of files that defines the pipeline in this case is:
 
-The following example shows a user-defined function:
+| File          | Description                                                                                         |
+|---------------|-----------------------------------------------------------------------------------------------------|
+| `program.sql` | The SQL code, including table, view, and function declarations.                                     |
+| `stubs.rs`    | Auto-generated by the SQL compiler; contains Rust stubs for UDFs ([see below](#step-2-locate-the-auto-generated-rust-prototype)).                       |
+| `udf.rs`      | User-provided implementation of UDFs in Rust.                                                       |
+| `udf.toml`    | List of external crates required by the code in `udf.rs` in the `Cargo.toml` format.                |
```

Comment:
User-provided list of external crates required by the code in `udf.rs` in the same format as that within the `[dependencies]` section of the well-known `Cargo.toml` format.